### PR TITLE
[processor/resourcedetectionprocessor] add k8snode detector

### PR DESCRIPTION
**Pull request by dloucasfx**
**Description:** <Describe what has changed.>
Adding k8snode detector to the resourcedetection processor.
We currently have no way to attach the k8s node UID attribute to logs/metrics/traces , this detector allows such an attribute to be added. 
Note: We can add more node attributes if needed, but atm only `k8s.node.uis` is added

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>
Unit test, mock tests and manual validation

**Documentation:** <Describe the documentation added.>
See Readme

### CODE REVIEW

**Comment by dmitryax** [comment]
@dloucasfx please rebase

File: processor/resourcedetectionprocessor/internal/k8snode/config.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	// NodeName is the actual name of the node we need to find the UID for.
+	// User can explicitly set it using the node_name option or using the env variable `K8S_NODE_NAME`
+	// In k8s deployment, you can assign the pod `fieldPath: spec.nodeName` to the env variable
+	NodeName           string                            `mapstructure:"node_name"`
+	ResourceAttributes metadata.ResourceAttributesConfig `mapstructure:"resource_attributes"`
+}
+
+// UpdateDefaults validates and update the default config with user's provided settings
+func (c *Config) UpdateDefaults() error {
+	if c.NodeName == "" {
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/internal/k8snode/config.go:25]: Do you think hardcoded the node name can be useful? I'd suggest instead making the env var name configurable (maybe with `K8S_NODE_NAME` being the default). See how it's done in k8s attributes receiver https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/k8sattributesprocessor/config.go#L244 .

File: processor/resourcedetectionprocessor/internal/k8snode/metadata.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+type: resourcedetectionprocessor/k8snode
+
+parent: resourcedetection
+
+resource_attributes:
+  k8s.node.uid:
+    description: The k8s.node.uid
```
[the remark this answers is not in the document]
Comment (dloucasfx) [reply processor/resourcedetectionprocessor/internal/k8snode/metadata.yaml]: PR updated with the changes

File: processor/resourcedetectionprocessor/internal/k8snode/config.go (hunk 1 above)
Comment (dloucasfx) [reply processor/resourcedetectionprocessor/internal/k8snode/config.go]: PR is updated with the changes

**Review by dmitryax — APPROVED** [review]

File: processor/resourcedetectionprocessor/internal/k8snode/config.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	ResourceAttributes metadata.ResourceAttributesConfig `mapstructure:"resource_attributes"`
+}
+
+// UpdateDefaults validates and update the default config with user's provided settings
+func (c *Config) UpdateDefaults() error {
+	if c.NodeFromEnvVar == "" {
+		c.NodeFromEnvVar = "K8S_NODE_NAME"
+	}
+	if value, envExists := os.LookupEnv(c.NodeFromEnvVar); !envExists || value == "" {
+		return fmt.Errorf("node name can't be found. Check the readme on how to set the requiered env variable")
```
Comment (mx-psi) [inline processor/resourcedetectionprocessor/internal/k8snode/config.go:46]:
```suggestion
		return fmt.Errorf("node name can't be found. Check the readme on how to set the required env variable")
```

**Review by mx-psi — APPROVED** [review]

**Comment by dmitryax** [comment]
@dloucasfx LGTM. Please rebase once more

File: internal/metadataproviders/k8snode/metadata.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+type Provider interface {
+	// NodeUID returns the K8S Node UID
+	NodeUID(ctx context.Context) (string, error)
+}
+
+type k8snodeProvider struct {
+	k8snodeClient kubernetes.Interface
+	nodeName      string
```
[the remark this answers is not in the document]
Comment (dloucasfx) [reply internal/metadataproviders/k8snode/metadata.go:23]: Yes, I will add it.
FYI, this noName is set by the caller, it is not necessarily the actual nodeName; but I can use the k8s API to confirm

**Review by mx-psi — APPROVED** [review]

File: processor/resourcedetectionprocessor/internal/k8snode/config_test.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	err := cfg.UpdateDefaults()
+	assert.NoError(t, err)
+}
+
+func TestSetCustomEnvVarEmptyVal(t *testing.T) {
+	cfg := CreateDefaultConfig()
+	cfg.NodeFromEnvVar = "my_env_var"
+	err := cfg.UpdateDefaults()
+	assert.Error(t, err)
+	assert.EqualError(t, err, "node name can't be found. Check the readme on how to set the requiered env variable")
```
Comment (codeboten) [inline processor/resourcedetectionprocessor/internal/k8snode/config_test.go:25]:
```suggestion
	assert.EqualError(t, err, "node name can't be found. Check the readme on how to set the required env variable")
```

Comment (dloucasfx) [reply processor/resourcedetectionprocessor/internal/k8snode/config_test.go]: thanks!

File: processor/resourcedetectionprocessor/README.md [hunk 6]
```diff
@@ -439,6 +439,61 @@ processors:
     override: false
 ```
 
+### K8S Node Metadata
+
+Queries the K8S api server to retrieve the following node resource attributes:
+
+    * k8s.node.uid
```
Comment (mx-psi) [inline processor/resourcedetectionprocessor/README.md:446]:
```suggestion
    * k8s.node.uid
    * k8s.node.name
```

File: .chloggen/resourcedetection_k8snode.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: processor/resourcedetectionprocessor
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: add k8snode detector to provide node metadata; currently the detector provides `k8d.node.uid`
```
Comment (mx-psi) [inline .chloggen/resourcedetection_k8snode.yaml:10]:
```suggestion
note: add k8snode detector to provide node metadata; currently the detector provides `k8d.node.uid` and `k8s.node.name`
```

**Comment by dmitryax** [comment]
@dloucasfx it needs another rebase :(

**Pull request merged**